impact_stack/rest/__init__.py

```diff
@@ -15,28 +16,48 @@
 class AuthMiddleware:
     """Requests middleware for authenticating using JWT tokens.
 
-    The middleware transparently requests an API-token from the auth-app on-demand.
-    """
+    The middleware transparently requests an API-token from the auth-app during the first request.
+    Afterwards each request gets the token added to its headers.
 
-    # pylint: disable=too-few-public-methods
-    # For now the middleware is very simple, but in the future it should also take care of caching
-    # the JWT and renew if needed.
+    A new token is requested on-demand whenever the old token has (almost) expired.
+    """
 
-    def __init__(self, client, api_key):
+    def __init__(self, client, api_key, token_life_time_buffer):
         """Create new auth-app requests auth middleware."""
         self.client = client
         self.api_key = api_key
+        self.life_time_buffer = token_life_time_buffer
+        self.token = None
+        self.expires_at = None
+
+    def needs_refresh(self):
+        """Check if we have a token and it can still be used."""
+        if not self.token:
+            return True
+        time_left = self.expires_at - utils.now().timestamp()
+        return time_left < self.life_time_buffer
 
     def get_token(self):
         """Use the API key to get a JWT."""
-        return self.client.post("token", json=self.api_key, json_response=True)["token"]
+        if self.needs_refresh():
+            data = self.client.post("token", json=self.api_key, json_response=True)
+            self.token = data["token"]
+            self.expires_at = data["data"]["exp"]
+        return self.token
 
     def __call__(self, request: requests.PreparedRequest):
         """Add the JWT token to the request."""
         request.headers["Authorization"] = "Bearer " + self.get_token()
         return request
 
 
+def timeout_sum(timeout):
+    """Sum up all the values in a requests timeout."""
+    # A timeout is a tuple of connect and read timeout. Passing an integer is a short hand for
+    # using the same number for both.
+    return timeout * 2 if isinstance(timeout, int) else sum(timeout)
```

Comment:
To me, `timeout_sum()` looks a bit lost here. Since it's only used by `ClientFactory`, it could be a method of that class, or if it’s a more general utility function, moved to utils?

Reply:
I have moved it into `ClientFactory` as suggested. I don’t have a strong opinion how this should look like.